hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -157,28 +171,29 @@ SqlResult execute(DropMappingPlan plan) {
     }
 
     SqlResult execute(CreateDataLinkPlan plan) {
-        InternalDataLinkService dlService = getNodeEngine(hazelcastInstance).getDataLinkService();
-        if (plan.ifNotExists() && dlService.existsDataLink(plan.name())) {
-            throw new HazelcastException("Data link '" + plan.name() + "' already exists");
-        }
+        InternalDataLinkService dlService = nodeEngine.getDataLinkService();
+        assert !plan.ifNotExists() || !plan.isReplace();
 
-        dlService.createSqlDataLink(plan.name(), plan.type(), plan.options(), plan.isReplace());
-        dataLinksCatalog.createDataLink(
-                new DataLink(plan.name(), plan.type(), plan.options()),
+        if (dlService.existsConfigDataLink(plan.name())) {
+            throw new HazelcastException("Cannot replace a data link created from configuration");
+        }
+        boolean added = dataLinksCatalog.createDataLink(
+                new DataLinkCatalogEntry(plan.name(), plan.type(), plan.options()),
                 plan.isReplace(),
                 plan.ifNotExists());
+        if (added) {
+            broadcastUpdateDataLinkOperations(plan.name());
```

Comment:
Hmm, shouldn't `broadcastUpdateDataLinkOperations` be under hood of `DataLinkService`? I wonder how config-based data links spreads in cluster, and if they are. If not, then it doesn't look consistent

Reply:
I designed it this way: the DLService only has responsibility for local `DataLink` objects. SQL uses the DLService to create DLs, so does dynamic config handler. But the dynamic config handler also distributes the config change operation to all members in its own code, and calls DLService on each member. SQL is analogous - the plan executor broadcasts the request to all members, and on each member creates the `DataLink` (not `DataLinkCatalogEntry`).

Reply:
Okay, it's clear now. Thanks